# WCDT_5hmc

Martin Sjöström, Shuang G. Zhao, Samuel Levy, Meng Zhang et al. Cancer Research 2022.

https://aacrjournals.org/cancerres/article/doi/10.1158/0008-5472.CAN-22-1123/709779/The-5-Hydroxymethylcytosine-Landscape-of-Prostate

## Abstract

Analysis of DNA methylation is a valuable tool to understand disease progression and is increasingly being used to create diagnostic and prognostic clinical biomarkers. While conversion of cytosine to 5-methylcytosine (5mC) commonly results in transcriptional repression, further conversion to 5-hydroxymethylcytosine (5hmC) is associated with transcriptional activation. Here we perform the first study integrating whole-genome 5hmC with DNA, 5mC, and transcriptome sequencing in clinical samples of benign, localized, and advanced prostate cancer. 5hmC is shown to mark activation of cancer drivers and downstream targets. Furthermore, 5hmC sequencing revealed profoundly altered cell states throughout the disease course, characterized by increased proliferation, oncogenic signaling, dedifferentiation, and lineage plasticity to neuroendocrine and gastrointestinal lineages. Finally, 5hmC sequencing of cell-free DNA from patients with metastatic disease proved useful as a prognostic biomarker able to identify an aggressive subtype of prostate cancer using the genes TOP2A and EZH2, previously only detectable by transcriptomic analysis of solid tumor biopsies. Overall, these findings reveal that 5hmC marks epigenomic activation in prostate cancer and identify hallmarks of prostate cancer progression with potential as biomarkers of aggressive disease.

Please direct correspondence related to this manuscript to Felix Feng, UCSF.  
Please diret inquiries related to the data in this manuscript to David Quigley, UCSF.

## Raw data

5hmC sequence data described in this study are available on the EGA website at dataset EGAD00001008462:

https://ega-archive.org/datasets/EGAD00001008462

Whole Genome Bisulfite Sequence data described in this study are available on the SRA repository:

https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=phs001648

Whole Genome DNA and RNA sequence data (previously reported) is available at the NCBI GDC repository:

https://portal.gdc.cancer.gov/projects/WCDT-MCRPC

Somatic variant calls related to this analysis are available at:

https://quigleylab.ucsf.edu/data

Scripts used for the results reported in this paper are available through this github website.

